### PR TITLE
Add Hash::from_inner and prevent double allocation in FromHex using it

### PR DESCRIPTION
**Pull request by stevenroose**

### CODE REVIEW

File: src/hex.rs [hunk 1]
```diff
@@ -84,6 +97,40 @@ impl<'a> Iterator for HexIterator<'a> {
...
+        let len = self.sl.len() / 2;
+        (len, Some(len))
+    }
+}
+
+impl<'a> DoubleEndedIterator for HexIterator<'a> {
+    fn next_back(&mut self) -> Option<Result<u8, Error>> {
+        let current_len = self.sl.len();
+        if current_len % 2 == 1 {
+            Some(Err(Error::OddLengthString(self.sl.len())))
```
Comment (stevenroose) [inline src/hex.rs]: Why not use `current_len` here?

File: src/hex.rs [hunk 2]
```diff
@@ -84,6 +97,40 @@ impl<'a> Iterator for HexIterator<'a> {
...
+        } else if self.sl.is_empty() {
```
Comment (stevenroose) [inline src/hex.rs]: Why not use `current_len` here as well? I would start with `if current_len == 0 { ... } else if current_len % 2 == 1 { ... }` personally.

Comment (stevenroose) [reply src/hex.rs]: But I see this is mirroring the impl of `Iterator`, which if fine I guess.

**Comment by stevenroose** [comment]
Can't approve my own PR. So ACK d71dd9959bd72626f245fdc6ed01d5b0a21db4db

File: src/hex.rs [hunk 3]
```diff
@@ -82,6 +97,40 @@ impl<'a> Iterator for HexIterator<'a> {
...
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.sl.len() / 2;
+        (len, Some(len))
+    }
+}
+
+impl<'a> DoubleEndedIterator for HexIterator<'a> {
+    fn next_back(&mut self) -> Option<Result<u8, Error>> {
+        let current_len = self.sl.len();
+        if current_len % 2 == 1 {
```
Comment (sgeisler) [inline src/hex.rs]: Imo this check shouldn't happen for every element, but I'll open a separate PR.

File: src/hex.rs (hunk 1 above)
Comment (sgeisler) [reply src/hex.rs]: I took the liberty to fix it in a separate commit, do you think I should squash it into Andrew's commit?

**Pull request merged**